File: testers/mariadb/scala/src/scala/testdb/withConnection.scala
package testdb

import dev.typr.foundationssc.*
import dev.typr.foundationssc.connect.{ConnectionSettings, MariaConfig, TransactionIsolation}

object withConnection {
  private val config = MariaConfig.builder("localhost", 3307, "typr", "typr", "password").build()
  private val settings = ConnectionSettings.builder().transactionIsolation(TransactionIsolation.READ_COMMITTED).build()
  private val transactor = Transactor.create(config, settings).rollbackOnly()

  def apply[T](f: Connection ?=> T): T = transactor.transact(f)
}

File: testers/mariadb/kotlin/src/kotlin/testdb/WithConnection.kt
package testdb

import dev.typr.foundations.connect.mariadb.MariaDbConfig
import java.sql.Connection

object WithConnection {
    private val CONFIG: MariaDbConfig =
        MariaDbConfig.builder("localhost", 3307, "typr", "typr", "password").build()

    fun <T> apply(f: (Connection) -> T): T {
        CONFIG.connect().use { conn ->
            conn.autoCommit = false
            try {
                return f(conn)
            } finally {
                conn.rollback()
            }
        }
    }

    fun run(f: (Connection) -> Unit) {
        CONFIG.connect().use { conn ->
            conn.autoCommit = false
            try {
                f(conn)
            } finally {
                conn.rollback()
            }
        }
    }
}
